Replicator/Pusher.cc

```diff
@@ -436,15 +450,40 @@ namespace litecore { namespace repl {
 
     // Called after a proposed revision gets a 409 Conflict response from the server.
     // Check the document's current remote rev, and retry if it's different now.
-    bool Pusher::shouldRetryConflictWithNewerAncestor(RevToSend *rev) {
-        // None of this is relevant if there's no puller getting stuff from the server
-        DebugAssert(_options.pull > kC4Passive);
-
+    bool Pusher::shouldRetryConflictWithNewerAncestor(RevToSend *rev, slice receivedRevID) {
         if (!_proposeChanges)
             return false;
         try {
             Retained<C4Document> doc = _db->getDoc(rev->docID, kDocGetAll);
             if (doc && C4Document::equalRevIDs(doc->revID(), rev->revID)) {
+                if(receivedRevID && receivedRevID != rev->remoteAncestorRevID) {
+                    // Remote ancestor received in proposeChanges response, so try with 
+                    // this one instead
+                    try {
+                        // If receivedRevID is newer than rev->revID this will throw since locally we don't have
+                        // that rev yet.  So the only way this function succeeds is if receivedRevID is equal to
+                        // or an ancestor of rev->revID
+                        if(doc->selectCommonAncestorRevision(rev->revID, receivedRevID)) {
```

Comment:
I don't like the idea of throwing an exception in normal operation. Why not preflight by testing `if (doc->selectRevision(receivedRevID))` instead?

Reply:
I like this idea much better.  I didn't like the idea of doing a try catch inside a try catch either....